cmd/server.go

```diff
@@ -145,45 +181,43 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
 					continue
 				}
 
-				// Try to take snapshot before setting
-				config := &miscellaneous.Config{
-					Attempts: 6,
-					Delay:    1,
-					Units:    time.Second,
-					Logger:   logger,
-				}
-				if err := miscellaneous.Do(func() error {
-					logger.Infof("Taking initial snapshot at time: %s", time.Now().Local())
-					err := ssr.TakeFullSnapshot()
-					if err != nil {
-						logger.Infof("Taking initial snapshot failed: %v", err)
-					}
-					return err
-				}, config); err != nil {
-					if etcdErr, ok := err.(*errors.EtcdError); ok == true {
-						logger.Errorf("Snapshotter failed with etcd error: %v", etcdErr)
-					} else {
-						logger.Fatalf("Snapshotter failed with error: %v", err)
-					}
-					handler.Status = http.StatusServiceUnavailable
-					continue
-				} else {
-					handler.Status = http.StatusOK
+				logger.Infof("Creating snapshotter...")
+				ssr := snapshotter.NewSnapshotter(
+					logger,
+					snapshotterConfig,
+					&ackCh)
+				if err != nil {
+					logger.Fatalf("Failed to create snapshotter from configured storage provider: %v", err)
 				}
 
+				go func() {
+					for {
+						select {
+						case ssrHealth, ok := <-ssr.HealthCh:
```

Comment:
Do we really need HealthCh?  I think, you can convert this go routine to simple function `UpdateHealth(bool)` And call where ever required.

Reply:
We need the health state of ssr in the handler. We are not passing the ssr object to the handler. If we need to update the handler of the health of the ssr object, we would either need to pass ssr to the handler or handler to the ssr object. Do you think thats a good idea? As for a global function, we cannot pass the ssr or handler instance in the same. How do you suggest we go about it?

Reply:
As we discussed, health status can be updated only in `server.go` before calling `ssr.Run` or after return from `ssr.Run`.  No need to update it from inside ssr routine.  Handler is just reading status field on /status request. So, no update call needed in handler. And since we are initializing handler in main thread of server command, there we can surely update the handler.status. If still not clear to you, lets sit together and refactor it.

Reply:
Are you suggesting we set the health status to OK and run the snapshotter. Change it when it fails? Earlier we could do it as we had the first snapshot run seperate. Now I have integrated that with ssr.Run(). I need to send out a response when the first snapshot succeeded to toggle the Ready state.